### PR TITLE
Implement IoTHubRegistryManager

### PR DESCRIPTION
**Pull request by zolvarga**
# Checklist
- [ ] I have read the [contribution guidelines] (https://github.com/Azure/azure-iot-sdk-node/blob/master/.github/CONTRIBUTING.md).
- [ ] I added or modified the existing tests to cover the change (we do not allow our test coverage to go down).
- If this is a modification that impacts the behavior of a public API
  - [ ] I edited the corresponding document in the `devdoc` folder and added or modified requirements.

# Reference/Link to the issue solved with this PR (if any)

# Description of the problem

# Description of the solution

### CODE REVIEW

File: service/src/cl/iothub_registry_manager.ts [hunk 1]
```diff
@@ -0,0 +1,968 @@
...
+   * @description                                             Creates a device identity on IoTHub using X509 authentication.
+   * @param {string}  deviceId                                The name (Id) of the device.
+   * @param {string}  eTag                                    The etag (if_match) value to use for the update operation.
+   * @param {string}  primaryThumbprint                       Primary X509 thumbprint.
+   * @param {string}  secondaryThumbprint                     Secondary X509 thumbprint.
+   * @param {boolean} isEnabled                               The status of the device. If the status disabled, a device cannot connect to the service.
+   * @returns Promise<DevicesCreateOrUpdateIdentityResponse>  The return object containing the updated device and the parsed HttpResponse.
+   * @memberof IoTHubRegistryManager
+   */
+  updateDeviceWithX509(
```
Comment (anthonyvercolano) [inline service/src/cl/iothub_registry_manager.ts:391]: >updateDeviceWithX509 [](start = 2, length = 20)

Why is this called update?  There is not capability for update if there is no device object.

Comment (anthonyvercolano) [reply service/src/cl/iothub_registry_manager.ts:391]: This and others with update*

---
In reply to: [571791877](https://github.com/Azure/azure-iot-sdk-node/pull/923#discussion_r571791877) [](ancestors = 571791877)

**Review by anthonyvercolano — CHANGES_REQUESTED** [review]
:clock1:

File: service/test/_iothub_registry_manager_test.js [hunk 2]
```diff
@@ -0,0 +1,689 @@
...
+  });
+
+  it('createDeviceWithSas calls the createOrUpdateIdentity method on the PL client', async () => {
+    var returnValue = {
+        interfaces: {
+          testInterfaceInstanceName: {}
+        }
+      };
+    var testDeviceId = 'testDeviceId';
+    var testPrimaryKey = 'testPrimaryKey';
```
Comment (anthonyvercolano) [inline service/test/_iothub_registry_manager_test.js:33]: >'testPrimaryKey' [](start = 25, length = 16)

I fear we will get dinged for hard coded keys in the code.  I know I have.  Substitute with a uuid?

File: service/test/_iothub_registry_manager_test.js [hunk 3]
```diff
@@ -0,0 +1,689 @@
...
+        sinon.match.has('ifMatch', testETag)
+    ));
+    assert.deepEqual(returnedPromise.interfaces, returnValue.interfaces);
+  });
+
+  it('invokeDeviceMethod calls the invokeMethod method on the PL client', async () => {
+    var returnValue = {
+        interfaces: {
+          testInterfaceInstanceName: {}
+        }
```
Comment (anthonyvercolano) [inline service/test/_iothub_registry_manager_test.js:654]: What are these?

File: service/src/cl/iothub_registry_manager.ts (hunk 1 above)
Comment (zolvarga) [reply service/src/cl/iothub_registry_manager.ts:391]: Fixed: updateDevice is introduced

File: service/test/_iothub_registry_manager_test.js (hunk 2 above)
Comment (zolvarga) [reply service/test/_iothub_registry_manager_test.js:33]: for the unit tests I don't think we need other than "a" value

File: service/test/_iothub_registry_manager_test.js (hunk 3 above)
Comment (zolvarga) [reply service/test/_iothub_registry_manager_test.js:654]: Artifical return values to verify.

File: service/src/cl/iothub_registry_manager.ts [hunk 4]
```diff
@@ -0,0 +1,779 @@
...
+
+    return this._pl.modules.createOrUpdateIdentity(deviceId, moduleId, module);
+  }
+
+  /**
+   * @method updateModule                                           module: azure-iot-digitaltwins-service.IoTHubRegistryManager.updateModule
+   * @description                                                   Updates a module identity for a device on IoTHub using SAS authentication.
+   * @param {string}  deviceId                                      The name (Id) of the device.
+   * @param {string}  moduleId                                      The name (Id) of the module.
+   * @param {Module}  device                                        The contents of the module identity.
```
Comment (anthonyvercolano) [inline service/src/cl/iothub_registry_manager.ts:492]: I think this should be module.

**Review by anthonyvercolano — APPROVED** [review]

**Comment by zolvarga** [comment]
azp run

**Pull request merged**